### PR TITLE
Support partition consumer receive async and fix batch logic

### LINKED ISSUE #36: Support partition-consumer receive async logic
**Is your feature request related to a problem? Please describe.**

Now, we not impl receive async logic of partition consumer.

**Describe the solution you'd like**

We should support receive async logic in partition consumer.

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <ranxiaolong716@gmail.com>

Fixes #36 

### Motivation

Support partition consumer receive async logic and fix batch logic for project.

Add test case:

- TestBatchMessageReceive
- TestConsumer_ReceiveAsync
- TestInvalidURL
- TestProducerConnectError
- TestProducerNoTopic
- TestEventTime
- TestFlushInProducer
- TestFlushInPartitionedProducer
- TestPartitionConsumerAckTimeout
- TestConsumer_Shared

### CODE REVIEW

**Comment by wolfstudy** [comment]
@merlimat @sijie @jiazhai PTAL thanks.

File: pulsar/impl_consumer.go [hunk 1]
```diff
@@ -154,27 +155,47 @@ func (c *consumer) Unsubscribe() error {
 }
 
 func (c *consumer) Receive(ctx context.Context) (Message, error) {
+	if len(c.consumers) > 1 {
+		for _, pc := range c.consumers {
+			go func(pc Consumer) {
+				for {
+					msg, err := pc.Receive(ctx)
```
Comment (jiazhai) [inline pulsar/impl_consumer.go]: Please follow java/cpp logic, load c.queue in background, not at the time receive is called.

File: pulsar/impl_consumer.go [hunk 2]
```diff
@@ -154,27 +155,47 @@ func (c *consumer) Unsubscribe() error {
...
+	}
+
+	return c.consumers[0].(*partitionConsumer).Receive(ctx)
+}
+
+func (c *consumer) ReceiveAsync(ctx context.Context, msgs chan<- ConsumerMessage) error {
 	for _, pc := range c.consumers {
 		go func(pc Consumer) {
-			if err := pc.ReceiveAsync(ctx, c.queue); err != nil {
+			if err := pc.ReceiveAsync(ctx, msgs); err != nil {
```
Comment (jiazhai) [inline pulsar/impl_consumer.go]: same as above comments

File: pulsar/impl_partition_consumer.go [hunk 3]
```diff
@@ -238,20 +238,55 @@ func (pc *partitionConsumer) internalUnsubscribe(unsub *handleUnsubscribe) {
...
 			if err != nil {
-				pc.log.WithError(err).Errorf("unserialize message id error:%s", err.Error())
 				return nil, err
 			}
-			if pc.unAckTracker != nil {
-				pc.unAckTracker.Add(id)
+
+			// increase available permits
+			receivedSinceFlow++
+			err = pc.increaseAvailablePermits(receivedSinceFlow)
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: please follow java implementatiaon, and put `trackMessage` and `increaseAvailablePermits` into one method called "messageProcessed",

File: pulsar/impl_partition_consumer.go [hunk 4]
```diff
@@ -274,30 +309,25 @@ func (pc *partitionConsumer) ReceiveAsync(ctx context.Context, msgs chan<- Consu
...
+
+				// increase available permits
 				receivedSinceFlow++
-				if receivedSinceFlow >= highwater {
-					if err := pc.internalFlow(receivedSinceFlow); err != nil {
-						pc.log.Errorf("Send Flow cmd error:%s", err.Error())
-						return err
-					}
-					receivedSinceFlow = 0
+				err = pc.increaseAvailablePermits(receivedSinceFlow)
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: same comments as above

File: pulsar/impl_partition_consumer.go [hunk 5]
```diff
@@ -274,30 +309,25 @@ func (pc *partitionConsumer) ReceiveAsync(ctx context.Context, msgs chan<- Consu
...
+				if err != nil {
+					return err
 				}
 				continue
 			}
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	}
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: Please add a receiveAsync method, that returns only one message, and following the implementation of Java client.

File: pulsar/impl_consumer.go [hunk 6]
```diff
@@ -154,27 +155,47 @@ func (c *consumer) Unsubscribe() error {
 }
 
 func (c *consumer) Receive(ctx context.Context) (Message, error) {
+	if len(c.consumers) > 1 {
```
Comment (jiazhai) [inline pulsar/impl_consumer.go]: In current pulsar master branch, it is already support partitionNumber == 1

File: pulsar/impl_consumer.go [hunk 7]
```diff
@@ -154,27 +155,47 @@ func (c *consumer) Unsubscribe() error {
...
+				return cMsg.Message, nil
+			}
+			return nil, errors.New("receive message error")
+		}
+	}
+
+	return c.consumers[0].(*partitionConsumer).Receive(ctx)
+}
+
+func (c *consumer) ReceiveAsync(ctx context.Context, msgs chan<- ConsumerMessage) error {
```
Comment (jiazhai) [inline pulsar/impl_consumer.go]: Please add a receiveAsync method, that returns only one message, and following the implementation of Java client.

File: pulsar/impl_partition_consumer.go [hunk 8]
```diff
@@ -580,50 +610,52 @@ func (pc *partitionConsumer) HandlerMessage(response *pb.CommandMessage, headers
 	id := newMessageID(int64(msgID.GetLedgerId()), int64(msgID.GetEntryId()),
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: Please change method name of `HandlerMessage` into "messageReceived" or "handleMessage"

File: pulsar/impl_partition_consumer.go [hunk 9]
```diff
@@ -580,50 +610,52 @@ func (pc *partitionConsumer) HandlerMessage(response *pb.CommandMessage, headers
...
-		var dup bool
-
-		pc.omu.Lock()
-		for _, mid := range pc.overflow {
-			if proto.Equal(mid, newMid) {
-				dup = true
-				break
+		select {
+		case pc.subQueue <- consumerMsg:
+			//Add messageId to Overflow buffer, avoiding duplicates.
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: why there would be duplicates?

Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: what is overflow buffer? why we need this?

File: pulsar/impl_consumer.go (hunk 6 above)
Comment (wolfstudy) [reply pulsar/impl_consumer.go]: In here, when `len(c.consumers) > 1` means the consumer is a `partitionConsumer`, when `len(c.consumers) = 1` means the consumer non partition.

File: pulsar/impl_partition_consumer.go [hunk 10]
```diff
@@ -274,30 +320,45 @@ func (pc *partitionConsumer) ReceiveAsync(ctx context.Context, msgs chan<- Consu
...
 			return ctx.Err()
 		}
 	}
+}
+
+func (pc *partitionConsumer) ReceiveAsyncWithCallback(ctx context.Context, callback func(msg Message, err error)) {
+	highwater := uint32(math.Max(float64(cap(pc.options.MessageChannel)/2), 1))
+
+	// request half the buffer's capacity
+	if err := pc.internalFlow(highwater); err != nil {
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: try to do it in background

File: pulsar/impl_partition_consumer.go [hunk 11]
```diff
@@ -238,32 +252,64 @@ func (pc *partitionConsumer) internalUnsubscribe(unsub *handleUnsubscribe) {
...
+	return message, err
 }
 
 func (pc *partitionConsumer) ReceiveAsync(ctx context.Context, msgs chan<- ConsumerMessage) error {
-	highwater := uint32(math.Max(float64(cap(pc.options.MessageChannel)/2), 1))
+	highWater := uint32(math.Max(float64(cap(pc.options.MessageChannel)/2), 1))
 
 	// request half the buffer's capacity
-	if err := pc.internalFlow(highwater); err != nil {
+	if err := pc.internalFlow(highWater); err != nil {
```
Comment (jiazhai) [inline pulsar/impl_partition_consumer.go]: also here, in background

**Pull request merged**